### PR TITLE
[Merged by Bors] - refactor(CategoryTheory/Monoidal): typeclasses Functor.LaxMonoidal, Functor.OplaxMonoidal and Functor.Monoidal

### PR DESCRIPTION
**Pull request by joelriou**
Monoidal functors are refactored. Given a functor `F : C ⥤ D` between monoidal categories, we introduce typeclasses `F.LaxMonoidal` and `F.OplaxMonoidal` which carry the data of morphisms like `μ : F.obj X ⊗ F.obj Y ⟶ F.obj (X ⊗ Y)` (or `δ : F.obj (X ⊗ Y) ⟶ F.obj X ⊗ F.obj Y`). Then, the functor is monoidal if both `F.LaxMonoidal` and `F.OplaxMonoidal` and that both data provide inverse isomorphisms. With this design, there is more symmetry between lax and oplax functors (in the bundled setting, this was also experimented upon in #10845 ). (Roughly speaking, what was previously in `CategoryTheory.Monoidal.Functorial` was generalized.)

This is also a change from the previous designs where (lax) monoidal functors were bundled, which would be quite problematic or at least very unpractical in future applications: sooner or later, we shall have functors between derived categories, which are lax/oplax/monoidal, but are also triangulated, and some of these functors shall be adjoints, etc. The bundled aspect was not very much in use before this refactor: it was relevant only when constructing functors from a category of lax monoidal functors; in such case, the corresponding case of bundled lax monoidal functors, etc, have been kept in the code.

---

- [x] depends on: #17761

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by joelriou** [comment]
!bench

File: Mathlib/CategoryTheory/Monoidal/Functor.lean [hunk 1]
```diff
@@ -35,719 +36,1038 @@ See <https://stacks.math.columbia.edu/tag/0FFL>.
...
+@[reassoc (attr := simp)]
+lemma whiskerRight_η_ε (T : D) : η F ▷ T ≫ ε F ▷ T = 𝟙 _ := by
+  rw [← MonoidalCategory.comp_whiskerRight, η_ε, id_whiskerRight]
+
+@[reassoc (attr := simp)]
+lemma whiskerRight_μ_δ (X Y : C) (T : D) : μ F X Y ▷ T ≫ δ F X Y▷ T = 𝟙 _ := by
+  rw [← MonoidalCategory.comp_whiskerRight, μ_δ, id_whiskerRight]
+
+@[reassoc (attr := simp)]
+lemma whiskerRight_δ_μ (X Y : C) (T : D) : δ F X Y ▷ T ≫ μ F X Y▷ T = 𝟙 _ := by
```
Comment (yuma-mizuno) [inline Mathlib/CategoryTheory/Monoidal/Functor.lean:437]:
```suggestion
lemma whiskerRight_δ_μ (X Y : C) (T : D) : δ F X Y ▷ T ≫ μ F X Y ▷ T = 𝟙 _ := by
```

File: Mathlib/CategoryTheory/Monoidal/Functor.lean [hunk 2]
```diff
@@ -35,719 +36,1038 @@ See <https://stacks.math.columbia.edu/tag/0FFL>.
...
+@[reassoc (attr := simp)]
+lemma whiskerRight_ε_η (T : D) : ε F ▷ T ≫ η F ▷ T = 𝟙 _ := by
+  rw [← MonoidalCategory.comp_whiskerRight, ε_η, id_whiskerRight]
+
+@[reassoc (attr := simp)]
+lemma whiskerRight_η_ε (T : D) : η F ▷ T ≫ ε F ▷ T = 𝟙 _ := by
+  rw [← MonoidalCategory.comp_whiskerRight, η_ε, id_whiskerRight]
+
+@[reassoc (attr := simp)]
+lemma whiskerRight_μ_δ (X Y : C) (T : D) : μ F X Y ▷ T ≫ δ F X Y▷ T = 𝟙 _ := by
```
Comment (yuma-mizuno) [inline Mathlib/CategoryTheory/Monoidal/Functor.lean:433]:
```suggestion
lemma whiskerRight_μ_δ (X Y : C) (T : D) : μ F X Y ▷ T ≫ δ F X Y ▷ T = 𝟙 _ := by
```

**Comment by joelriou** [comment]
!bench

**Comment by joelriou** [comment]
I do not know why `Algebra.Homology.HomotopyCategory` has become significantly slower (I will try to investigate), but `CategoryTheory.Triangulated.Opposite` was an extremely slow file, any improvement on this is good!

**Comment by joelriou** [comment]
!bench

**Comment by joelriou** [comment]
!bench

**Review by TwoFX — APPROVED** [review]
I think this is a good refactor and the changes look good to me.

Could you merge master once more before sending the PR to bors, just to make sure no semantic merge conflict was added since you last merged master?

bors d+

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**